### PR TITLE
feat: [CDS-42833]: Make possible to disable user .gitconfig in Git delegate commands

### PR DESCRIPTION
**Pull request by vcalasansh**
## Describe your changes
This PR adds a boolean option `disableUserGitConfig` to GitConfig class (the spec of Git connectors configuration in Harness CG) which allows disabling the configurations in the `.gitconfig` file located in the filesystem's user directory.

Also adds a feature flag `SPG_ALLOW_DISABLE_USER_GITCONFIG` for UI to selectively show this option for Git connectors.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38085)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/38085)

File: 960-api-services/src/main/java/io/harness/git/GitClientV2Impl.java [hunk 1]
```diff
@@ -1359,4 +1372,14 @@ TransportCommand getAuthConfiguredCommand(TransportCommand gitCommand, GitBaseRe
     }
     return gitCommand;
   }
+
+  private void clearUserConfigIfDisabled(Boolean disableUserConfig) {
+    if (disableUserConfig != null && disableUserConfig) {
+      try {
+        SystemReader.getInstance().getUserConfig().clear();
```
Comment (abhinav-harness) [inline 960-api-services/src/main/java/io/harness/git/GitClientV2Impl.java:1379]: will this remove from system permanently ?

File: 400-rest/src/test/java/software/wings/service/impl/yaml/GitSyncServiceImplTest.java [hunk 2]
```diff
@@ -92,6 +92,7 @@ public void setup() {
                                           .urlType(GitConfig.UrlType.REPO)
                                           .authenticationScheme(AuthenticationScheme.HTTP_PASSWORD)
                                           .branch("branchName")
+                                          .disableUserGitConfig(false)
```
Comment (abhinav-harness) [inline 400-rest/src/test/java/software/wings/service/impl/yaml/GitSyncServiceImplTest.java:95]: I am not sure but do we want to do it other way - which means enable user config and then set it as true and override user value in jgit?

File: 930-delegate-tasks/src/main/java/software/wings/service/impl/GitServiceImpl.java [hunk 3]
```diff
@@ -98,6 +98,7 @@ public void ensureRepoLocallyClonedAndUpdated(GitOperationContext gitOperationCo
                                                       .authRequest(authRequest)
                                                       .accountId(gitConfig.getAccountId())
                                                       .connectorId(gitOperationContext.getGitConnectorId())
+                                                      .disableUserGitConfig(gitConfig.getDisableUserGitConfig())
```
Comment (abhinav-harness) [inline 930-delegate-tasks/src/main/java/software/wings/service/impl/GitServiceImpl.java:101]: need gql and yaml changes for this param

**Review by vcalasansh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38085)* status: 21 file(s) to review. 3. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/test/java/software/wings/service/impl/yaml/GitSyncServiceImplTest.java` line 95 at r1](https://reviewable.io/reviews/harness/harness-core/38085#-ND53IjM4k8N5A7EXtiB-r1-95:-NFEfU0u-OzQJs8xpySs:bjw9u9) ([raw file](https://github.com/harness/harness-core/blob/d510ebd22abcadffba2f82385c87206310d8b4e9/400-rest/src/test/java/software/wings/service/impl/yaml/GitSyncServiceImplTest.java#L95)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

I am not sure but do we want to do it other way - which means enable user config and then set it as true and override user value in jgit?
</blockquote></details>

Thank you for the comment! I would think it makes more sense this way, since using the user's gitconfig is the standard and default behavior of Git. But if you think its best the other way around let me know, I can easily change this.
___
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/GitServiceImpl.java` line 101 at r1](https://reviewable.io/reviews/harness/harness-core/38085#-ND4farZ1e18PbuQFdw8-r1-101:-NFEfs3I4_Jr0E3oJ6LU:bol3kou) ([raw file](https://github.com/harness/harness-core/blob/d510ebd22abcadffba2f82385c87206310d8b4e9/930-delegate-tasks/src/main/java/software/wings/service/impl/GitServiceImpl.java#L101)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

need gql and yaml changes for this param
</blockquote></details>

Done! Thank you!
___
*[`960-api-services/src/main/java/io/harness/git/GitClientV2Impl.java` line 1379 at r1](https://reviewable.io/reviews/harness/harness-core/38085#-ND4farZ1e18PbuQFdwB-r1-1379:-NFEfvZq0zLpinbZU3qe:b-hcruij) ([raw file](https://github.com/harness/harness-core/blob/d510ebd22abcadffba2f82385c87206310d8b4e9/960-api-services/src/main/java/io/harness/git/GitClientV2Impl.java#L1379)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

will this remove from system permanently ?
</blockquote></details>

Update the PR to do this with a custom instance of SystemReader. Thank you!

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> Reviewed 11 of 17 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38085)* status: 10 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/test/java/software/wings/service/impl/yaml/GitSyncServiceImplTest.java` line 95 at r1](https://reviewable.io/reviews/harness/harness-core/38085#-ND53IjM4k8N5A7EXtiB-r1-95:-NFGPsw281bcQ3D-AGpJ:ba698zx) ([raw file](https://github.com/harness/harness-core/blob/d510ebd22abcadffba2f82385c87206310d8b4e9/400-rest/src/test/java/software/wings/service/impl/yaml/GitSyncServiceImplTest.java#L95)):*
<details><summary><i>Previously, vcalasansh (Vinícius Calasans) wrote…</i></summary><blockquote>

Thank you for the comment! I would think it makes more sense this way, since using the user's gitconfig is the standard and default behavior of Git. But if you think its best the other way around let me know, I can easily change this.
</blockquote></details>

Yes it does make change - but i will not make a behavioral change for a customer unless we need it as they can be using this "hack" without us knowing it.
___
*[`930-delegate-tasks/src/main/java/software/wings/service/impl/yaml/GitClientImpl.java` line 1341 at r2](https://reviewable.io/reviews/harness/harness-core/38085#-NFGQAyZ8CuFl2KFpG6p:-NFGQAyZ8CuFl2KFpG6q:bdgdjyk) ([raw file](https://github.com/harness/harness-core/blob/f7ab80576f65a874b464f1a47490940071a08fe2/930-delegate-tasks/src/main/java/software/wings/service/impl/yaml/GitClientImpl.java#L1341)):*
> ```Java
>       } catch (Exception ex) {
>         log.error("Exception in loading user GitConfig: ", ex);
>       }
> ```

do we need some finally too?

**Review by abhinav-harness — DISMISSED** [review]
> Reviewed 4 of 8 files at r1, 4 of 17 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38085)* status: :shipit: complete! ✓

**Review by abhinav-harness — DISMISSED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38085)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38085)* status: :shipit: complete! ✓

**Comment by vcalasansh** [comment]
trigger codebasehash

**Pull request merged**